### ISSUE TITLE
Implement Proof-of-Concept

### ISSUE THREAD

**Issue by hbeni** [hbeni/fgcom-mumble#1]
We need a PoC; it should be developed with the following features:

- [x] **Prio 1: basic mumble plugin**
  Basic mumble plugin will allow basic pilot radio operation already
  - [x] Write basic plugin (see https://github.com/mumble-voip/mumble/pull/3743; full specification at [plugin.spec.md](https://github.com/hbeni/fgcom-mumble/blob/master/plugin.spec.md))
     - [x] basic plugin functions/code setup/layout
     - [x] UDP input interface
     - [x] mute/unmute based on selected frequencies (range done <s>may follow shortly later, if its complicated to calculate?</s>)
  - [x] `echo "..." | netcat -q1 -u localhost 16661` really is good enough for now. A standalone radio client may be useful in the future tough! followup=Issue #4 <s>implement a very basic PTT/Radio sim client so we can test the plugin easily (a simple UDP command line wrapper tool may be sufficient, and in fact useful for later scripting purposes... simply netcat with some checking logic ontop?)</s>

----
- [x] **Prio 2: design and conduct load tests**
Basic mumble load and bandwith requirements are the same, but we should get some test for our usecase (pilot/atc radio situation), as i guess, it is not realistic that _n_-clients speak and listen concurrently for 100% of the time... But also with constantly broadcasting ATIS stations...
  - [x] Develop load testing infrastructure (custom bots<s>? some existing load test tool? stock or adapted [mumble benchmark](https://github.com/mumble-voip/mumble/blob/master/src/tests/Benchmark.cpp)?</s>)
  - [x] We should make sure that the usability, responsiveness and bandwith-size with higher client counts stays within acceptable limits.
  - [x] We should try to get some numbers for maximum supported clients; this is a main indicator if this project has a future

----
- [x] **Prio 3: example FGFS and ATC integration**
  see [plugin.spec.md#plugin-input-data](https://github.com/hbeni/fgcom-mumble/blob/master/plugin.spec.md#plugin-input-data)
  - [x] Flightgear compatibility
    - [x] <s>ensure basic compatibility to current fgcom fgfs protocol (working with default fgcom keybinds and c172p)
        Needs change to core; for the problems see: https://forum.flightgear.org/viewtopic.php?p=370172#p370172</s>
    - [x] implement the new protocol with flightgear (working with c172 and c182s)
  - [x] ATC-Pie compatibility
  - [x] openRadar compatibility

----
- [x] **Prio 4: basic ATIS recording support**
  see [README.md#the-fgcom-mumble-client-plugin](https://github.com/hbeni/fgcom-mumble/blob/master/README.md#the-fgcom-mumble-client-plugin)
  - [x] write simple recording bot
  - [x] write simple playback bot
  - [x] make ATIS cycle work (record -> playback)
  - [x] make 910.00 work

**Comment by hbeni** [comment]
Bots are currently missing implementation of plugin data sending:
https://github.com/bkacjios/lua-mumble/issues/13

**Comment by hbeni** [comment]
> Bots are currently missing implementation of plugin data sending:
> [bkacjios/lua-mumble#13](https://github.com/bkacjios/lua-mumble/issues/13)

Was fixed yesterday and now bis wie like a charm.
Simple echo bot already working, now coding proper recording and playback bots.
Load Testing could be done relatively easy using specialized Play-back bots, which are easily formed off once Play-back bot is alive.

We are turning final on this as it looks!

**Comment by hbeni** [comment]
ATIS / 910.00 is working now! There still is a bug however with invocing the playback bots, they somehow block the recorder from functioning. it does work if the recorder just records (--nospawn option) and then to manually invoke playback bots.

A workaround, if this is not possible, would be to have some shell script wrapper (or lua one) that checks for new playback files and the starting the playback bots from the shell.

**Comment by hbeni** [comment]
ATC-Pie and openRadard already work with default old FGCom protocol (you need one instance of mumble for each radio).
ATC-Pie provides some advanced codes that use the newer FGCom-mumble protocol. This needs to be tested still.

**Comment by hbeni** [comment]
> A workaround, if this is not possible, would be to have some shell script wrapper (or lua one) that checks for new playback files and the starting the playback bots from the shell.

- [x] Implemented it, now we have a bot manager.

**Comment by hbeni** [comment]
- Load testing with spawning radio-playback bots transmitting a previously recorded ATIS sample looped showed that **50 clients** (in LAN here) and then conducting an Echotest recording yields acceptable results.
- Trying with a special implemented fake-pilot bot is currently not possible because lua eats itself with more than 10 bot instances. With those numbers, there is no problem at all.

So it looks currently like assumed: the mumble backend together with network hardware infrastructure is the main limiting factor.

**Comment by hbeni** [comment]
Further testing: I can reproduce this without any special code from me.

The following fails here after about 100 invocations.

First the loop call code (bash script):
```bash
function callTest {
    i=$1; echo "CALL=$i"
    openssl genrsa -out /tmp/bot-$i.key 2048 2> /dev/null
    openssl req -new -sha256 -key /tmp/bot-$i.key -out /tmp/bot-$i.csr -subj "/"
    openssl x509 -req -in /tmp/bot-$i.csr -signkey /tmp/bot-$i.key -out /tmp/bot-$i.pem 2> /dev/null

    lua testTimer.lua --id=$i --host=localhost --port=64738 --cert=/tmp/bot-$i.pem --key=/tmp/bot-$i.key  1>/dev/null 2>/dev/null &
}

for i in {1..100}; do
  callTest $i
  sleep 0.5
done
```

```lua
mumble = require("mumble")

-- Parse cmdline args
local host   = "localhost"
local port   = 64738
local cert   = "bot.pem"
local key    = "bot.key"
local secs   = 0.5  -- seconds for timer
local botid  = ""

if arg[1] then
    if arg[1]=="-h" or arg[1]=="--help" then
        print("testTimer bot")
        print("usage: "..arg[0].." [opt=val ...]")
        print("  Options:")
        print("    --id=      id to join with              (default=random)")
        print("    --host=    host to connect to           (default="..host..")")
        print("    --port=    port to connect to           (default="..port..")")
        print("    --cert=    path to PEM encoded cert     (default="..cert..")")
        print("    --key=     path to the certs key        (default="..key..")")
        print("    --secs=    Seconds between timer loops  (default="..secs..")")
        os.exit(0)
    end
    
    for _, opt in ipairs(arg) do
        _, _, k, v = string.find(opt, "--(%w+)=(.+)")
        if k=="id"      then botid=v end
        if k=="host"    then host=v end
        if k=="port"    then port=v end
        if k=="cert"    then cert=v end
        if k=="key"     then key=v end
        if k=="secs"    then secs=v end
    end
    
end

-- parameter checks
--if sample == "" then print("parameter --sample is mandatory!") os.exit(1) end

callsign = "PLUGINSPAM-"
if botid == "" then
    callsign = callsign..math.random(1, 99999)
else
    callsign = callsign..botid
end




-- Connect to server, so we get the API
print("connecting as '"..callsign.."' to "..host.." on port "..port.." (cert: "..cert.."; key: "..key..")")
local client = assert(mumble.connect(host, port, cert, key))
client:auth(callsign)
print("connect and bind: OK")

-- timer loop
local checkTimer = mumble.timer()
timerCode = function(t)
    print("tick")
end

client:hook("OnServerSync", function(event)
    print("Sync done; server greeted with: ", event.welcome_text)
    
    -- start timer   
    checkTimer:start(timerCode, 0.00, secs)
   
end)



mumble.loop()
print("callsign "..callsign.." completed.")
```

**Comment by hbeni** [comment]
ATC-Pie (the development version) successfully tested. Transmitting/receiving on different radios did work, even the shared radio sending.
ATIS recording/playback (using the lua bots) did work too.

**Comment by hbeni** [comment]
> The load test bots fail after some higher number instantiation currently. This may be a problem in the mumble-lua lib or in luajit directly. Its currently investigated at [bkacjios/lua-mumble#14 (comment)](https://github.com/bkacjios/lua-mumble/issues/14#issuecomment-654763196))

The problem was mitigated by using `luajit` instead of `lua`.
Conducted a load test with the following counts:
- 50 fake pilot bots flying around and chatted randomly (25% chance that they play a 5s sample every 30 secs; 10% chance that this sample is an echotest (so more than in reallife)).
- 5 ATIS recordings broadcasting
- The recorder bot responding to random 910.00 echotests from the bots
I picked a time slot whee the radio bot currently was recording for another bot to record my 910.00 testsample. This worked fine. the sample got recorded, the replay bot spawned and he echoed it back to me. My voice was loud and clear as expected.

**So to summarize**, with 50 pilots chatting and 5 parallel ATIS messages, there seem to be no problem with the plugin infrastructure. Sounds good to me.

**Comment by hbeni** [comment]
bkacjios did another fix at the lua library, which enabled me to spawn about 100 bots.
No problem at that load, however, my client machine slowly comes to an edge where i can't test more.
